sklearn_numba_dpex/common/random.py

```diff
@@ -0,0 +1,309 @@
+import warnings
+import random
+from functools import lru_cache
+
+from numba import float32, float64, uint32, int64, uint64
+import numpy as np
+
+import dpctl
+import dpctl.tensor as dpt
+import numba_dpex as dpex
+
+# This code is largely inspired from the numba.cuda.random module and the
+# numba/cuda/random.py where it's defined (v<0.57), and by the implementation of the
+# same algorithm in the package `randomgen`.
+
+# numba.cuda.random: https://github.com/numba/numba/blob/0.56.3/numba/cuda/random.py
+# randomgen: https://github.com/bashtage/randomgen/blob/v1.23.1/randomgen/xoroshiro128.pyx
+
+# NB1: we implement xoroshiro128++ rather than just xoroshiro128+, which is preferred.
+# Reference resource about PRNG: https://prng.di.unimi.it/
+
+# NB2: original numba.cuda.random code also includes functions for generating normally
+# distributed floats but we don't include it here as long as it's not needed.
+
+zero_idx = int64(0)
+one_idx = int64(1)
+
+
+def get_random_raw(states):
+    """Returns one `uint64` random integer"""
+    result = dpt.empty(sh=(1,), dtype=np.uint64, device=states.device)
+    make_random_raw_kernel()(states, result)
+    return result
+
+
+@lru_cache
+def make_random_raw_kernel():
+    @dpex.kernel
+    def _get_random_raw_kernel(states, result):
+        result[zero_idx] = _xoroshiro128pp_next(states, zero_idx)
+
+    return _get_random_raw_kernel[1, 1]
+
+
+@lru_cache
+def make_rand_uniform_kernel_func(dtype):
+    """Instanciate a kernel function that returns a random float in [0, 1)"""
+    if not hasattr(dtype, "name"):
+        raise ValueError(
+            "dtype is expected to have an attribute 'name', like np.dtype or numba "
+            "types."
+        )
+
+    if dtype.name == "float64":
+        convert_rshift = uint32(11)
+        convert_const = float64(uint64(1) << uint32(53))
+        convert_const_one = float64(1)
+
+        @dpex.func
+        def uint64_to_unit_float(x):
+            """Convert uint64 to float64 value in the range [0.0, 1.0)"""
+            return float64(x >> convert_rshift) * (convert_const_one / convert_const)
+
+    elif dtype.name == "float32":
+        convert_rshift = uint32(40)
+        convert_const = float32(uint32(1) << uint32(24))
+        convert_const_one = float32(1)
+
+        @dpex.func
+        def uint64_to_unit_float(x):
+            """Convert uint64 to float32 value in the range [0.0, 1.0)
+            NB: this is different than original numba.cuda.random code. Instead of
+            generating a float64 random number before casting it to float32, a float32
+            number is generated from uint64 without intermediate float64. This enables
+            compatibility with devices that do not support float64 numbers.
+            However is seems to be exactly equivalent e.g it passes the float precision
+            test in sklearn.
+            """
```

Comment:
PEP257 nitpick.

```suggestion
            """Convert uint64 to float32 value in the range [0.0, 1.0)

            NB: this is different than original numba.cuda.random code. Instead of
            generating a float64 random number before casting it to float32, a float32
            number is generated from uint64 without intermediate float64. This enables
            compatibility with devices that do not support float64 numbers.
            However is seems to be exactly equivalent e.g it passes the float precision
            test in sklearn.
            """
```